lib/models/branch.js

```diff
@@ -1,21 +1,86 @@
-const DETACHED = {};
+const DETACHED = Symbol('detached');
+const REMOTE_TRACKING = Symbol('remote-tracking');
 
 export default class Branch {
-  constructor(name, detached = null) {
+  constructor(name, upstream = nullBranch, push = upstream, head = false, attributes = {}) {
     this.name = name;
-    this.detached = detached === DETACHED;
+    this.upstream = upstream;
+    this.push = push;
+    this.head = head;
+    this.attributes = attributes;
   }
 
   static createDetached(describe) {
-    return new Branch(describe, DETACHED);
+    return new Branch(describe, nullBranch, nullBranch, true, {[DETACHED]: true});
+  }
+
+  static createRemoteTracking(refName, remoteName, remoteRef) {
+    return new Branch(name, nullBranch, nullBranch, false, {[REMOTE_TRACKING]: {remoteName, remoteRef}});
   }
 
   getName() {
     return this.name;
   }
 
+  getFullRef() {
+    if (this.isDetached()) {
+      return '';
+    }
+
+    if (this.isRemoteTracking()) {
+      if (this.name.startsWith('refs/')) {
+        return this.name;
+      } else if (this.name.startsWith('remotes/')) {
+        return `refs/${this.name}`;
+      }
+      return `refs/remotes/${this.name}`;
+    }
+
+    if (this.name.startsWith('refs/')) {
+      return this.name;
+    } else if (this.name.startsWith('heads/')) {
+      return `refs/${this.name}`;
+    } else {
+      return `refs/heads/${this.name}`;
+    }
+  }
+
+  getRemoteName() {
+    if (!this.isRemoteTracking()) {
+      return Symbol('not-remote-tracking');
```

Comment:
> Do we want them all to be unique and be referentially non-identical?

This is exactly what I was going for. I wanted the result of calling `.getRemoteName()` and the other conditional getters on a Branch to which they don't apply to never compare `===` to the result from any other Branch, valid or invalid. If I returned `''` or a shared Symbol, this could happen:

```js
// Detached branch
const b0 = Branch.detached('HEAD~2');

// Branch with no upstream
const b1 = new Branch('untracked');

if (b0.getRemoteName() === b1.getRemoteName()) {
  console.log('oops');
}
```

Mostly, I wanted to avoid needing to do a bunch of if guards any time I wanted to compare Branches in PrSelectionByBranch - in general, callers would need to know which of `isPresent()`, `isDetached()`, `isRemoteTracking()` to check on the operands before you checked the results of any given getter.

The _downside_ of doing this - in addition to looking a little weird - is that now you can no longer display `getRemoteName()` nicely in UI without having to do an `isRemoteTracking()` check first. It's also inconsistent with the way other null object getters behave (which tend to return something like `''` for uniform display). I'd convinced myself yesterday that this was worth it to streamline the comparisons I was doing, but now I'm not so sure. What do you think?

Reply:
Ah yes, makes sense. I think either way - no super strong feelings, so I leave it up to you 😄. 

I guess one guiding question could be: which do you expect to be more likely - comparing branches or displaying branch names in the UI? I did a quick check throughout the codebase and didn't seem to find any branch comparisons, but perhaps I just missed them.

Reply:
> I did a quick check throughout the codebase and didn't seem to find any branch comparisons, but perhaps I just missed them.

... Nope, that's because there aren't any any more! I think this is another case where I'd designed a bit of the model layer with one revision of the container code in mind, and ended up implementing the container in a different way.

> I guess one guiding question could be: which do you expect to be more likely - comparing branches or displaying branch names in the UI?

With that in mind, this answer definitely becomes "displaying branch names in the UI." Modified accordingly 👍